shared/Microsoft.Extensions.Certificates.Generation.Sources/CertificateManager.cs

```diff
@@ -0,0 +1,414 @@
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+
+namespace Microsoft.Extensions.Certificates.Generation
+{
+    internal class CertificateManager
+    {
+        public const string AspNetHttpsOid = "1.3.6.1.4.1.311.84.1.1";
+        public const string AspNetHttpsOidFriendlyName = "Asp.Net Core HTTPS development certificate";
+
+        public const string AspNetIdentityOid = "1.3.6.1.4.1.311.84.1.2";
+        public const string AspNetIdentityOidFriendlyName = "Asp.Net Core Identity Json Web Token signing development certificate";
+
+        private const string ServerAuthenticationEnhancedKeyUsageOid = "1.3.6.1.5.5.7.3.1";
+        private const string ServerAuthenticationEnhancedKeyUsageOidFriendlyName = "Server Authentication";
+
+        private const string LocalhostHttpsDnsName = "localhost";
+        private const string LocalhostHttpsDistinguishedName = "CN=" + LocalhostHttpsDnsName;
+
+        private const string IdentityDistinguishedName = "CN=Microsoft.AspNetCore.Identity.Signing";
+
+        public const int RSAMinimumKeySizeInBits = 2048;
+
+        public IList<X509Certificate2> ListCertificates(
+            CertificatePurpose purpose,
+            StoreName storeName,
+            StoreLocation location,
+            bool isValid)
+        {
+            var certificates = new List<X509Certificate2>();
+            try
+            {
+                using (var store = new X509Store(storeName, location))
+                {
+                    store.Open(OpenFlags.ReadOnly);
+                    certificates.AddRange(store.Certificates.OfType<X509Certificate2>());
+                    IEnumerable<X509Certificate2> matchingCertificates = certificates;
+                    switch (purpose)
+                    {
+                        case CertificatePurpose.All:
+                            matchingCertificates = matchingCertificates
+                                .Where(c => HasOid(c, AspNetHttpsOid) || HasOid(c, AspNetIdentityOid))
+                                .ToList();
+                            break;
+                        case CertificatePurpose.HTTPS:
+                            matchingCertificates = matchingCertificates
+                                .Where(c => HasOid(c, AspNetHttpsOid))
+                                .ToList();
+                            break;
+                        case CertificatePurpose.Signing:
+                            matchingCertificates = matchingCertificates
+                                .Where(c => HasOid(c, AspNetIdentityOid))
+                                .ToList();
+                            break;
+                        default:
+                            break;
+                    }
+                    if (isValid)
+                    {
+                        // Ensure the certificate hasn't expired, has a private key and its exportable
```

Comment:
Why check for exportable? What if someone doesn't want autogenerated and creates their own non-exportable key?

Reply:
Non exportable certificates can't be used by docker